### DISCUSSION TITLE
How does ergo compare?

### DISCUSSION THREAD

**Discussion by heri16** [ergo-services/ergo#107] [Q&A]
How does ergo compare to this similar library in terms of performance and developer experience?

https://proto.actor/

https://github.com/asynkron/protoactor-go#why-actors

**Comment by halturin** [comment]
I'm not super familiar with this framework. 

The only pro I see in this framework is using lock-free MPSC instead of channels for the actor's communications. It makes communicating a way more performant. I already have a task in my roadmap to implement communication the same way in the future.

The cons - it doesn't provide a native way for interop with Erlang/Elixir cluster, it doesn't have a Proxy feature. I also don't see anything like Monitor/Link features. It doesn't check the incarnation of the process. I haven't found a way how to make sync requests.

I should also mention that they use GRPC, which is a pro for some cases but could be a con for others.

Anyway, it's a good framework. I suppose I miss something and would appreciate it if anyone else could make a way more expressive comparison.

**Reply by qjpcpu** [reply]
@heri16 @halturin 
This is a very late comment, haha...

I've actually been using proto-actor for years, but I'm now planning to move to Ergo.

Proto-actor is great for getting started, but you start to lose precise control as your project gets more complex. For instance:

* It uses the virtual actor model. Placing actors is cheap, but with a huge number of them, load balancing is tough to manage.

* The supervisors are very lightweight, so I found myself needing a more precise supervision tree.

* The source code is a bit messy, and the API doesn't feel production-ready. I'm stuck on a forked old version, and upgrading is a pain.

* It pulls in a lot of dependencies, and upgrading is a big job because of library conflicts. Ergo having zero dependencies is really cool.

> I've gone through all of Ergo's code and documentation. I noticed the big refactor in 2024, which makes me a little concerned about stability—that's when I happened to come across this discussion.

Still, proto-actor is a solid choice for someone just getting into the actor model or for pure Go devs.

**Comment by Northward712** [comment]
Protobuf is a double-edged sword for protoactor-go. While it provides a core advantage, it also acts as its primary limitation, especially when dealing with types that are not natively supported by Protobuf.
